syne_tune/experiments/plotting.py

```diff
@@ -434,11 +435,28 @@ def _figure_shape(plot_params: PlotParameters) -> Tuple[int, int]:
             nrows = ncols = 1
         return nrows, ncols
 
+    def _transform_dataframe(
+        self,
+        df: pd.DataFrame,
+        plot_params: PlotParameters,
+        dataframe_column_generator: Optional[DataFrameColumnGenerator],
+    ) -> pd.DataFrame:
+        if dataframe_column_generator is None:
+            return df
+        metric = plot_params.metric
+        assert metric not in df.columns, (
+            f"New column to be appended to results dataframe: {metric} is "
+            f"already a column: {df.columns}"
+        )
+        new_column = dataframe_column_generator(df)
+        return df.assign(**{metric: new_column})
+
     def _aggregrate_results(
         self,
         df: pd.DataFrame,
         plot_params: PlotParameters,
         extra_results_keys: Optional[List[str]],
+        dataframe_column_generator: Optional[DataFrameColumnGenerator],
```

Comment:
I think this should be extracted from `_aggregate_results` to separate aggregation and transformation. I believe it is important to have one function/module fulfil one purpose as much as we can (see single responsibility principle). In this case, I think `plot` can accept the transformation given its current design but we should separate aggregation and enrichement. In fact `plot` should first update the setup_df using something like (can be inside the transform function with extra asserts):
```
setup_df["new column"] = None
for tuner_name, tuner_df in setup_df.groupby("tuner_name"):
    setup_df.loc[tuner_df.index, "new column"] = dataframe_column_generator(tuner_df)
                
```

and then `_aggregate` should not worry about any transformations anymore. This way both functions can be modified/extended independently as they do not share any concerns.

Reply:
I disagree. First of all, these are some internal methods.
Second, there is grouping happening first w.r.t. setup, then w.r.t. tuner (i.e., seed). Then, you do 2 things on the dataframes grouped this way: add column, then aggregate.

A copy&paste job doing this 2-level grouping in two methods is not only inefficient, but also not clean.

I can rename the method, and introduce another helper method for the aggregation only. But let us remain realistic.

Reply:
My latest refactoring should address the concerns

Reply:
My feeling is you get a bit worked up on a pretty isolated point. In much of our code, lots of different things happen in a loop.

I can now pick any of these codes, and ask for two or three loops. What does this add?

We have something like this

Grouping loop:
- Do thing_a (by calling a method called thing_a)
- Do thing b on output of thing_a (by calling a method called thing_b)

You ask me to rewrite this into:

do_thing_a_with_grouping:
- Grouping loop:
   - Do thing_a
   
 do_thing_b_with_grouping:
 - Grouping loop:
    - Do thing_b
    
I really fail to see why the latter is better design.

It is not the case that slower code means better design.

Reply:
I know this single responsibility idea. My code does that. But as methods become higher-level, their single responsibility becomes more complex.

It is important to note a few things:
- The column to add must be computed from the grouped data for a single experiment only
- The dataframe with the extra column is not maintained, it is just an intermediate for the aggregation
- I've been adding extra methods to separate concerns

Reply:
I am also not sure if you looked at the most recent code. The method is not called `_aggregate_results` anymore.